### PR TITLE
Lanczos tests

### PR DESCRIPTION
**Pull request by aprokop**

### CODE REVIEW

File: include/mfmg/dealii/amge_host.templates.hpp [hunk 1]
```diff
@@ -192,7 +192,8 @@ AMGe_host<dim, MeshEvaluator, VectorType>::compute_local_eigenvectors(
   {
     boost::property_tree::ptree lanczos_params;
     lanczos_params.put("num_eigenpairs", n_eigenvectors);
-    lanczos_params.put("tolerance", tolerance);
+    // lanczos_params.put("tolerance", tolerance);
+    lanczos_params.put("tolerance", 1e-6);
```
Comment (dalg24) [inline include/mfmg/dealii/amge_host.templates.hpp]: Did you mean to commit that?

File: tests/test_hierarchy.cc [hunk 2]
```diff
@@ -73,9 +73,10 @@ double test(std::shared_ptr<boost::property_tree::ptree> params)
       laplace._dof_handler, laplace._constraints, a, material_property);
   mfmg::Hierarchy<DVector> hierarchy(comm, evaluator, params);
 
-  pcout << "Grid complexity    : " << hierarchy.grid_complexity() << std::endl;
-  pcout << "Operator complexity: " << hierarchy.operator_complexity()
-        << std::endl;
+  // pcout << "Grid complexity    : " << hierarchy.grid_complexity() <<
+  // std::endl; pcout << "Operator complexity: " <<
+  // hierarchy.operator_complexity()
+  // << std::endl;
```
Comment (dalg24) [inline tests/test_hierarchy.cc]: Remove do not comment it out.

What is the rational for having/not having these print to the standard output?

File: tests/test_hierarchy.cc [hunk 3]
```diff
@@ -193,35 +195,51 @@ BOOST_DATA_TEST_CASE(
...
-    else
-      BOOST_TEST(
-          conv_rate ==
-              ref_solution[std::make_tuple(mesh, distort_random, reordering)],
-          tt::tolerance(1e-6));
+    //   (geometry, distortion, reordering, eigensolver, matrix-free)
+    std::map<std::tuple<std::string, bool, std::string, std::string, bool>,
+             double>
+        ref_solution;
+    // clang-format off
```
Comment (dalg24) [inline tests/test_hierarchy.cc]: Readability: Good call !

File: tests/test_hierarchy.cc [hunk 4]
```diff
@@ -193,35 +195,51 @@ BOOST_DATA_TEST_CASE(
...
-              ref_solution[std::make_tuple(mesh, distort_random, reordering)],
-          tt::tolerance(1e-6));
-    else
-      BOOST_TEST(
-          conv_rate ==
-              ref_solution[std::make_tuple(mesh, distort_random, reordering)],
-          tt::tolerance(1e-6));
+    //   (geometry, distortion, reordering, eigensolver, matrix-free)
+    std::map<std::tuple<std::string, bool, std::string, std::string, bool>,
+             double>
```
Comment (dalg24) [inline tests/test_hierarchy.cc]: Readability: I would prefer a string or a scoped enumeration to denote matrix-free mode.

File: tests/test_hierarchy.cc [hunk 5]
```diff
@@ -193,35 +195,51 @@ BOOST_DATA_TEST_CASE(
...
+    ref_solution[std::make_tuple("hyper_cube" , true  , "Reverse Cuthill_McKee" , "lanczos" , false)] = 0.0488968016;
+    ref_solution[std::make_tuple("hyper_cube" , true  , "Reverse Cuthill_McKee" , "lanczos" , true)]  = 0.1575200977;
+    ref_solution[std::make_tuple("hyper_ball" , false , "None"                  , "lanczos" , false)] = 0.0953745739;
+    ref_solution[std::make_tuple("hyper_ball" , false , "None"                  , "lanczos" , true)]  = 0.3017812295;
+    ref_solution[std::make_tuple("hyper_ball" , false , "Reverse Cuthill_McKee" , "lanczos" , false)] = 0.0953745739;
+    ref_solution[std::make_tuple("hyper_ball" , false , "Reverse Cuthill_McKee" , "lanczos" , true)]  = 0.3017812295;
+    ref_solution[std::make_tuple("hyper_ball" , true  , "None"                  , "lanczos" , false)] = 0.0966058675;
+    ref_solution[std::make_tuple("hyper_ball" , true  , "None"                  , "lanczos" , true)]  = 0.2982453175;
+    ref_solution[std::make_tuple("hyper_ball" , true  , "Reverse Cuthill_McKee" , "lanczos" , false)] = 0.0966058675;
+    ref_solution[std::make_tuple("hyper_ball" , true  , "Reverse Cuthill_McKee" , "lanczos" , true)]  = 0.2982453175;
```
Comment (dalg24) [inline tests/test_hierarchy.cc]: TL;DR
Would you confirm that you did add convergences rates for Lanczos and that nothing else change?  Can you comment on the new gold values?  Are they reasonably close to their LAPACK's counterpart?

File: tests/test_hierarchy.cc [hunk 6]
```diff
@@ -193,35 +195,51 @@ BOOST_DATA_TEST_CASE(
...
-    if (mesh == std::string("hyper_cube"))
-      BOOST_TEST(
-          conv_rate ==
-              ref_solution[std::make_tuple(mesh, distort_random, reordering)],
-          tt::tolerance(1e-6));
-    else
-      BOOST_TEST(
-          conv_rate ==
-              ref_solution[std::make_tuple(mesh, distort_random, reordering)],
-          tt::tolerance(1e-6));
```
Comment (dalg24) [inline tests/test_hierarchy.cc]: Can you comment here?  Did we have `if (mesh is hyper cube) { expression; } else { same expression; }`?

File: tests/test_restriction_matrix.cc [hunk 7]
```diff
@@ -334,6 +340,6 @@ BOOST_AUTO_TEST_CASE(weight_sum, *utf::tolerance(1e-4))
       e[i] = 1.;
     e.compress(::dealii::VectorOperation::insert);
     restrictor_matrix.vmult(ee, e);
-    BOOST_TEST(ee.l1_norm() == 1.);
+    BOOST_TEST(ee.l1_norm() == 1., tt::tolerance(1e-4));
```
Comment (dalg24) [inline tests/test_restriction_matrix.cc]: What is the rational for having the floating point comparison manipulation in the `BOOST_TEST` macro versus having a decorator for the unit test?

File: include/mfmg/dealii/dealii_utils.hpp [hunk 8]
```diff
@@ -79,6 +79,9 @@ void matrix_market_output_file(
     const std::string &filename,
     const dealii::TrilinosWrappers::SparseMatrix &matrix);
 
+void matrix_market_output_file(const std::string &filename,
+                               const dealii::SparseMatrix<double> &matrix);
```
Comment (dalg24) [inline include/mfmg/dealii/dealii_utils.hpp]: Please comment on how this gets called with the changes you propose?

**Comment by dalg24** [comment]
Related to #115 and https://github.com/ORNL-CEES/mfmg/issues/114#issuecomment-464282096

File: include/mfmg/dealii/dealii_utils.hpp (hunk 8 above)
Comment (aprokop) [reply include/mfmg/dealii/dealii_utils.hpp]: It does not, should have been a separate PR for that.

File: include/mfmg/dealii/amge_host.templates.hpp (hunk 1 above)
Comment (aprokop) [reply include/mfmg/dealii/amge_host.templates.hpp]: Yes, but this is all in progress. I need to put a lot more comments about why I'm doing things this way.

File: tests/test_hierarchy.cc (hunk 2 above)
Comment (aprokop) [reply tests/test_hierarchy.cc]: I'll remove them.

File: tests/test_hierarchy.cc (hunk 5 above)
Comment (aprokop) [reply tests/test_hierarchy.cc]: Yes. The values are really close but not sure how robust the thing is. Lanczos test now fails, so I may need to play around with things. The whole thing is annoying, but proper fix is unclear.

File: tests/test_hierarchy.cc (hunk 6 above)
Comment (aprokop) [reply tests/test_hierarchy.cc]: yes

File: tests/test_restriction_matrix.cc (hunk 7 above)
Comment (aprokop) [reply tests/test_restriction_matrix.cc]: I was not sure that `BOOST_AUTO_TEST_CASE` and `BOOST_DATA_TEST_CASE` behave the same way, and was too lazy to google. So just replaced with what I know.

File: include/mfmg/dealii/amge_host.templates.hpp (hunk 1 above)
Comment (dalg24) [reply include/mfmg/dealii/amge_host.templates.hpp]: Please do so.

File: tests/test_hierarchy.cc (hunk 2 above)
Comment (dalg24) [reply tests/test_hierarchy.cc]: Check with @Rombur.  I don't have an opinion on this and I am not sure why it was here in the 1st place.

File: include/mfmg/dealii/dealii_utils.hpp (hunk 8 above)
Comment (dalg24) [reply include/mfmg/dealii/dealii_utils.hpp]: Please remove from that PR.

**Comment by aprokop** [comment]
Superseded by #136.

**Pull request closed without merging**